Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,717 @@
+consolidate functionality. To learn more about how to create custom rules, the
+document, [Bazel
+extensions](https://docs.bazel.build/versions/master/skylark/concepts.html)
+contains a comprehensive overview.
+
+### Macros
+
+Like [rules](#Rules) and [aspects](#Aspects), macros are defined in `.bzl`
+files.  A macro is a convenient way to call a rule, and not recognized by Bazel
+in the same way a rule is. As Bazel builds targets, instantiated by rules, it can't
+"build" a macro. A macro is evaluated during the evaluation of a BUILD file and
+instantiates rules. Once the macro is translated into a rule, the macro is
+irrelevant to the build process. In C++ compilation, this behavior is similar
+to the pre-processor feature of the compiler. Please see the coming example.
+
+_Note: The main distinction between a `.bzl` and a `BUILD` file is `BUILD` files are
+used to create targets by calling macros and rules. `.bzl` files define the
+implementation._
+
+Most users of Bazel implement a higher level system of macros to encapsulate
+defaults of building libraries and simplify configuration management. To create
+a wrapper for `objc_library`, create the file `objc_library.bzl`. The following
+macro restricts the customization, and enforces defaults of the native
+`objc_library` rule.
+
+
+```
+def objc_library(name, srcs=[], hdrs=[], deps=[], data=[]):
+    """
+    An objc_library that turns on pedantic warnings,
+    and enables modules
+    """
+    native.objc_library(
+        name=name,
+        deps=deps,
+        data=data,
+        srcs=srcs,
+        hdrs=hdrs,
+        enable_modules=True,
+        copts=[
+            "-Wpedantic"
+        ]
+    )
+```
+
+By loading our `objc_library` into a `BUILD` file, it will override the native
+`objc_library` rule, which is automatically imported. This is ideal because it
+prevents users from accidentally using `native.objc_library` when they should
+be using the one that prevents overriding `copts`. This pattern is also used in
+[`rules_python`](https://github.com/bazelbuild/rules_python).
+
+```
+load(":objc_library.bzl", "objc_library")
+
+# This calls the custom objc_library macro. Thankfully, building this with
+# Bazel will error out due to the unavailable argument `copts` used here.
+objc_library(name="some", copts=["-DSOME"])
+```
+
+The same principles can be applied to many rules: wrapping macros with macros.
+Please see the
+[`objc_library`](https://docs.bazel.build/versions/master/be/objective-c.html)
+documentation for all possible arguments. _Note: Unlike Starlark rules which
+are easily added on to Bazel, the `objc_library` is part of the internal Java
+rules shipped with the Bazel
+binary._ 
+
+_In the Xcode world, configuration is defined by Xcode projects and proprietary
+xcconfig. In Bazel it's easy to establish conventions and enforce consistency
+with abstractions like `objc_library`, and refactor_
```

Comment:
How does the Pinterest build do the .bzl to .xcconfig (or if it is the other way around I don't know)? We have both filetypes in our repository. Thanks.

Reply:
The original usage and unfortunate linage of xcconfig for diagnostics is documented on github:
https://github.com/pinterest/xchammer/blob/master/Docs/FastAndReproducibleBuildsWithXCHammer.md#diagnostics

Simply put, when we migrated to Bazel we generated the `.bzl` and manually synced them up when the diags change - change the `.xcconfig`, change the `.bzl`

The program mentioned in the doc is included with xchammer now and devs have used this in the past when they refactored diagnostic settings. These diagnostic settings don't often have a reason to change
```
tools/XCConfigExporter:xcconfig-exporter
```

Ideally, someone can propose a solution to disable defaults and sidestep this problem. Otherwise, I image an xcconfig evaluator could be integrated as a repository rule or toolchain, I'd be open to reviewing a proposal for this but would prefer the latter.

Reply:
I opened an issue here, perhaps we can resolve this in the future: https://github.com/pinterest/xchammer/issues/245

Reply:
So I suppose what I'm hoping for is a "You need to change a flag? Here is how you do it." type of entry. This document seems to be more: "You want to set up bazel for iOS? Here you go!". However, I think we need: "So we have a bazel build, but it uses some other specific things found on github. You need to do X? Here is how you do that with PINS current set up." I see a use for this document for external developers, but I think we need more for PINS developers.